tripal/tripal_jbrowse
# Integration with JBrowse2 by providing a config.json via URL

Issue #63 & #67 

The previous milestone on 4.x embedded a JBrowse2 instance by providing the configuration files (assemblies and tracks, separately) locally at a specific filepath, but there was not a way to provide these needed files when creating the instance through the interface. This PR sets out to do that, but with the common single configuration file (ex: config.json) that is normally associated with JBrowse2 instances. 

Specific tasks implemented:
1. The previous "JBrowse URL" field that prompted for a JBrowse instance's full URL was converted to a URL field for its configuration file instead. Commit f9822e5c254cb95e2969b265a9a92a396a387df6
2. The "JBrowse version" field has been removed, since this field implied that JBrowse 1 is currently supported. Since the JBrowse1 is no longer being actively maintained, we aren't putting in the effort at this time to test and ensure it is supported. Commit e309dd23a67f79a2846e842fd27764c4620dcd4f
3. A new field has been provided for setting the default start location for the embedded JBrowse2. The default behaviour is that the user will be prompted to select a chromosome before seeing the linear genome view. Commit e309dd23a67f79a2846e842fd27764c4620dcd4f
4. The 2 fields mentioned in 1) and 3) are now being provided to the application by being added as variables in drupalSettings and then accessing them using AJAX. Commit bdb4f91a6c795c5649568a13abdabc8ca407bac8 and e309dd23a67f79a2846e842fd27764c4620dcd4f

### Example
The embedded JBrowse2 page: <img width="1187" alt="Screenshot 2024-04-18 at 6 19 17 PM" src="https://github.com/tripal/tripal_jbrowse/assets/7927760/76d113cd-db14-4103-a410-a405303f0684">

The form:
<img width="681" alt="Screenshot 2024-04-18 at 6 20 53 PM" src="https://github.com/tripal/tripal_jbrowse/assets/7927760/bff0a736-ccfc-42ec-98c4-b103ea88daed">


**NOTES:** 
- A thorough tutorial on how to utilize this functionality is in progress and will be in a separate PR. 
- This functionality is currently limited to displaying a linear genome view for one assembly at a time

### Testing
1. Clone this repository and switch to the branch `tv4-renderJBrowseFromURL`
2. Build the docker using the following commands:
``` 
docker build . --tag=tripal_jbrowse:latest
docker run -dit --name=tripaljbrowse --publish=80:80 --volume=`pwd`:/var/www/drupal/web/modules/contrib/tripal_jbrowse tripal_jbrowse:latest
docker exec tripaljbrowse service postgresql restart
```
3. Go to http://localhost and log in using the credentials in the [Tripal Docker docs](https://tripaldoc.readthedocs.io/en/latest/install/docker.html#development-site-information)
4. Go to `Administration > Tripal > Content > +Add Tripal Content` to manually create an Organism and Genome Assembly
6. Go back to the Tripal content page and click on `+Add jbrowse instance` under the tab `JBrowse Instances`.
7. Fill out the `JBrowse Configuration URL` field by:
   - Providing a valid URL to a configuration file in JSON format. 
   - If you do not have a configuration file hosted anywhere, you can borrow one from the `test_data/` folder in the [JBrowse2 repo](https://github.com/GMOD/jbrowse-components). Copy it into your clone of this repository (For ex: `cp config.json tripal_jbrowse/`). Now check the file exists at the following URL, assuming your docker is on port 80: http://localhost/modules/contrib/tripal_jbrowse/config.json
   - Note that your config.json will need to pull data for reference sequence and tracks from valid URLs, otherwise the files need to also be hosted within your cloned repo.
8. Set a default location using the format `[chr][start pos]..[end pos]` or a the name of a feature in a feature track of your config.json.
9. Select the Organism and Genome Assembly you created in step 4 from their respective dropdowns in the form
10. Also test by leaving the 2 new fields blank. Expected behaviour: 
  - The JBrowse Configuration URL field should complain about not being left empty, and should not let you save the form.
  - The Default Start Location field can be left empty, and your JBrowse instance should look similar to this:
  
<img width="756" alt="Screenshot 2024-04-18 at 6 18 16 PM" src="https://github.com/tripal/tripal_jbrowse/assets/7927760/c8a40170-a006-4a37-a68e-9f2da43d71ed">
